src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        /// Headers that are interesting (i.e may extend our consensus tip) will marked to download there full blocks.
+        /// </summary>
+        /// <remarks>
+        /// The headers are assumed to be consecutive in order. 
+        /// </remarks>
+        /// <param name="networkPeerId">The network id that is presenting this headers.</param>
+        /// <param name="headers">The list of headers to connect to the chain tree.</param>
+        /// <returns>Indicators of what blocks need to be downloaded.</returns>
+        public ConnectedHeaders ConnectNewHeaders(int networkPeerId, List<BlockHeader> headers)
+        {
+            Guard.NotNull(headers, nameof(headers));
+            this.logger.LogTrace("({0}:{1},{2}.{3}:{4})", nameof(networkPeerId), networkPeerId, nameof(headers), nameof(headers.Count), headers.Count);
+
+            if (!this.chainedHeadersByHash.ContainsKey(headers[0].HashPrevBlock))
+            {
+                this.logger.LogTrace("(-)[HEADER_COULD_NOT_CONNECT]");
+                throw new ConnectHeaderException();
+            }
+
+            List<ChainedHeader> newChainedHeaders = this.CreateNewHeaders(headers);
+
+            this.SetPeerTip(networkPeerId, headers.Last().GetHash());
+
+            if (newChainedHeaders == null)
+            {
+                this.logger.LogTrace("(-)[NO_NEW_HEADER]");
+                return new ConnectedHeaders();
+            }
+
+            ChainedHeader earliestNewHeader = newChainedHeaders.First();
+            ChainedHeader latestNewHeader = newChainedHeaders.Last();
+
+            ConnectedHeaders connectedHeaders = null;
+
+            bool isAssumedValidEnabled = this.consensusSettings.BlockAssumedValid != null;
+            bool isWithinCheckpoints = this.consensusSettings.UseCheckpoints && (earliestNewHeader.Height <= this.checkpoints.GetLastCheckpointHeight()); ;
+
+            if (isWithinCheckpoints || isAssumedValidEnabled)
+            {
+                ChainedHeader currentChainedHeader = latestNewHeader;
+
+                while (currentChainedHeader != earliestNewHeader)
+                {
+                    if (currentChainedHeader.HashBlock == this.consensusSettings.BlockAssumedValid)
+                    {
+                        this.logger.LogDebug("Chained header '{0}' represents an assumed valid block.", currentChainedHeader);
+
+                        connectedHeaders = this.HandleAssumedValidHeader(currentChainedHeader, latestNewHeader);
+                        break;
+                    }
+
+                    CheckpointInfo checkpoint = this.checkpoints.GetCheckpoint(currentChainedHeader.Height);
+                    if (checkpoint != null)
+                    {
+                        this.logger.LogDebug("Chained header '{0}' is a checkpoint.", currentChainedHeader);
+
+                        connectedHeaders = this.HandleCheckpointsHeader(currentChainedHeader, latestNewHeader, checkpoint);
```

Comment:
So basically here, before calling the method, we should save download To if there is any from assume valid case and after calling the method, we should set it back to the old value as it will be always better than the one returned from this method. If we don't have such a value then we just work with what this method gives us obviously.

Reply:
Same seems to be needed for downloadfrom.